mm2src/coins/z_coin.rs

```diff
@@ -380,6 +429,240 @@ impl ZCoin {
         sync_guard.respawn_guard.watch_for_tx(tx.txid());
         Ok(tx)
     }
+
+    async fn tx_history_from_sql(
+        &self,
+        limit: usize,
+        paging_options: PagingOptionsEnum<i64>,
+    ) -> Result<SqlTxHistoryRes, MmError<SqlTxHistoryError>> {
+        let wallet_db = self.z_fields.light_wallet_db.clone();
+        async_blocking(move || {
+            let db_guard = wallet_db.lock();
+            let conn = db_guard.sql_conn();
+
+            let total_sql = SqlBuilder::select_from(TRANSACTIONS_TABLE)
+                .field("COUNT(id_tx)")
+                .sql()
+                .expect("valid SQL");
+            let total_tx_count = conn.query_row(&total_sql, NO_PARAMS, |row| row.get(0))?;
+
+            let mut sql_builder = SqlBuilder::select_from(name!(TRANSACTIONS_TABLE; "txes"));
+            sql_builder
+                .field("txes.txid")
+                .field("txes.id_tx as internal_id")
+                .field("txes.block as block");
+
+            let offset = match paging_options {
+                PagingOptionsEnum::PageNumber(page) => (page.get() - 1) * limit,
+                PagingOptionsEnum::FromId(id) => {
+                    offset_by_id(conn, &sql_builder, [id], "id_tx", "block DESC, id_tx ASC", "id_tx = ?1")?
+                        .ok_or(SqlTxHistoryError::FromIdDoesNotExist(id))?
+                },
+            };
+
+            let sql = sql_builder
+                .field("blocks.time")
+                .field("COALESCE(rn.received_amount, 0)")
+                .field("COALESCE(sn.sent_amount, 0)")
+                .left()
+                .join("(SELECT tx, SUM(value) as received_amount FROM received_notes GROUP BY tx) as rn")
+                .on("txes.id_tx = rn.tx")
+                // detecting spent amount by "spent" field in received_notes table
+                .join("(SELECT spent, SUM(value) as sent_amount FROM received_notes GROUP BY spent) as sn")
+                .on("txes.id_tx = sn.spent")
+                .join(BLOCKS_TABLE)
+                .on("txes.block = blocks.height")
+                .group_by("internal_id")
+                .order_by("block", true)
+                .order_by("internal_id", false)
+                .offset(offset)
+                .limit(limit)
+                .sql()
+                .expect("valid query");
+
+            let sql_items = conn
+                .prepare(&sql)?
+                .query_map(NO_PARAMS, |row| {
+                    let mut tx_hash: Vec<u8> = row.get(0)?;
+                    tx_hash.reverse();
+                    Ok(ZCoinSqlTxHistoryItem {
+                        tx_hash,
+                        internal_id: row.get(1)?,
+                        height: row.get(2)?,
+                        timestamp: row.get(3)?,
+                        received_amount: row.get(4)?,
+                        spent_amount: row.get(5)?,
+                    })
```

Comment:
Maybe this can be made on to a function
`fn sql_tx_item_from_row(row: &Row<'_>) -> Result<ZCoinSqlTxHistoryItem, SqlError>` to simplify `tx_history_from_sql` a little.

Reply:
Will do 🙂